x/dex/keeper/limit_order_book.go

```diff
@@ -0,0 +1,255 @@
+}
+
+func (k Keeper) GetLimitOrderBook0To1(
+	ctx context.Context,
+	pair *types.TradingPair,
+	tick *types.Tick,
+) *LimitOrderBook {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	priceMakerToTaker := tick.Price0To1
+	priceTakerToMaker := sdk.OneDec().Quo(*priceMakerToTaker)
+
+	fillIndex := tick.LimitOrderTranche0To1.FillTrancheIndex
+	placeIndex := tick.LimitOrderTranche0To1.PlaceTrancheIndex
+
+	token0, _ := PairToTokens(pair.PairId)
+	fillTrancheDTO, found := k.GetLimitOrderTranche(
+		sdkCtx,
+		pair.PairId,
+		tick.TickIndex,
+		token0,
+		fillIndex,
+	)
+	if !found {
+		panic("There should be a tranche here!")
+	}
+	fillTranche := NewLimitOrderTranche(
+		tick,
+		&fillTrancheDTO,
+		*priceMakerToTaker,
+		priceTakerToMaker,
+	)
+
+	placeTranche := fillTranche
+	if fillIndex != placeIndex {
+		placeTrancheDTO, found := k.GetLimitOrderTranche(
+			sdkCtx,
+			pair.PairId,
+			tick.TickIndex,
+			token0,
+			placeIndex,
+		)
+		if !found {
+			panic("There should be a tranche here!")
+		}
+		placeTranche = NewLimitOrderTranche(
+			tick,
+			&placeTrancheDTO,
+			*priceMakerToTaker,
+			priceTakerToMaker,
+		)
+	}
+
+	return NewLimitOrderBook0To1(
+		pair,
+		tick,
+		fillTranche,
+		placeTranche,
+	)
+}
+
+func NewLimitOrderBook1To0(
+	pair *types.TradingPair,
+	tick *types.Tick,
+	fillTranche *LimitOrderTranche,
+	placeTranche *LimitOrderTranche,
+) *LimitOrderBook {
+	return &LimitOrderBook{
+		Pair:                pair,
+		Tick:                tick,
+		LimitTrancheIndexes: tick.LimitOrderTranche1To0,
+		Tranches: map[uint64]*LimitOrderTranche{
+			tick.LimitOrderTranche1To0.FillTrancheIndex:  fillTranche,
+			tick.LimitOrderTranche1To0.PlaceTrancheIndex: placeTranche,
+		},
+	}
+}
+
+func NewLimitOrderBook0To1(
+	pair *types.TradingPair,
+	tick *types.Tick,
+	fillTranche *LimitOrderTranche,
+	placeTranche *LimitOrderTranche,
+) *LimitOrderBook {
+	return &LimitOrderBook{
+		Pair:                pair,
+		Tick:                tick,
+		LimitTrancheIndexes: tick.LimitOrderTranche0To1,
+		Tranches: map[uint64]*LimitOrderTranche{
+			tick.LimitOrderTranche0To1.FillTrancheIndex:  fillTranche,
+			tick.LimitOrderTranche0To1.PlaceTrancheIndex: placeTranche,
+		},
+	}
+}
+
+func (l *LimitOrderBook) NewTranche(trancheIndex uint64) *LimitOrderTranche {
+	token0, _ := PairToTokens(l.Pair.PairId)
+	var priceMakerToTaker, priceTakerToMaker sdk.Dec
+	if l.TokenIn == token0 {
+		priceMakerToTaker = *l.Tick.Price0To1
+	} else {
+		priceTakerToMaker = sdk.OneDec().Quo(*l.Tick.Price0To1)
+	}
+	return NewLimitOrderTranche(
+		l.Tick,
+		&types.LimitOrderTranche{
+			PairId:           l.Pair.PairId,
+			TokenIn:          l.TokenIn,
+			TickIndex:        l.Tick.TickIndex,
+			TrancheIndex:     trancheIndex,
+			ReservesTokenIn:  sdk.ZeroInt(),
+			ReservesTokenOut: sdk.ZeroInt(),
+			TotalTokenIn:     sdk.ZeroInt(),
+			TotalTokenOut:    sdk.ZeroInt(),
+		},
+		priceMakerToTaker,
+		priceTakerToMaker,
+	)
+}
+
+func (l LimitOrderBook) Swap(maxAmount sdk.Int) (
```

Comment:
Hmm. I think if anything the top-level should be changed, but we're moving away from a strictly procedural architecture where function names are unique to one where function names are tightly associated with a specific type.

Reply:
Understood that at the liquidity level they will have the same name. But IMHO if we are moving towards more polymorphism not reusing function names in cases where they aren't interchangeable becomes even more important.

Reply:
Trying to follow the double negative... I agree that reusing function names in case where they are interchangeable is important, that's why I'm calling it Swap.